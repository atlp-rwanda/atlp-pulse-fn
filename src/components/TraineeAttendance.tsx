/* eslint-disable */
import React, { useState } from 'react';
import { useTranslation } from 'react-i18next';
import { useQuery } from '@apollo/client';
import Pagination from '../components/Pagination';
import AttendanceData from '../dummyData/attendance.json';
import useDocumentTitle from '../hook/useDocumentTitle';
import { GET_WEEKLY_ATTENDANCE } from '../Mutations/Attendance';

interface TraineeAttendanceData {
  getTraineeAttendanceByID: {
    weekNumber: string;
    traineeAttendance: {
      days: string;
      value: number;
    }[];
  }[];
}

interface weekData {
  weekNumber: string;
  traineeAttendance: (
    | { days: string; value: number }
    | { days: string; value: string }
  )[];
}
[];

export function optimizeWeekData(weeklyData: TraineeAttendanceData) {
  const weekDays = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday'];

  const optimizedWeekData = weeklyData.getTraineeAttendanceByID.map((week) => {
    const daysPresent = week.traineeAttendance.map((day) => day.days);
    const missingDays = weekDays.filter((day) => !daysPresent.includes(day));
    const missingDayData = missingDays.map((day) => ({
      days: day,
      value: 'N/A',
    }));

    return {
      weekNumber: week.weekNumber,
      traineeAttendance: [...week.traineeAttendance, ...missingDayData],
    };
  });
  return optimizedWeekData;
}

export function getTraineeEmail() {
  const localAuth = localStorage.getItem('auth');
  if (localAuth) {
    const auth = JSON.parse(localAuth);
    return auth.email;
  }
}

const TraineeAttendance: React.FC = () => {
  useDocumentTitle('Attendance');
  const { t } = useTranslation();
  const [weekAttendance, setWeekAttendance] = useState<any>([]);
  const { loading, error, data } = useQuery<TraineeAttendanceData>(
    GET_WEEKLY_ATTENDANCE,
    {
      variables: { traineeEmail: getTraineeEmail() },
      onCompleted: (data) => {
        const optimizedWeekData = optimizeWeekData(data);
        setWeekAttendance(optimizedWeekData);
      },
      onError: (error: any) => {
        console.log(error);
      },
    },
  );

  if (error) {
    return <p>Error loading data</p>;
  }

  const {
    firstContentIndex,
    lastContentIndex,
    nextPage,
    prevPage,
    page,
    gaps,
    setPage,
    totalPages,
  } = Pagination({
    contentPerPage: 3,
    count: AttendanceData.length,
  });

  return (
    <>
      <div className="bg-light-bg dark:bg-dark-frame-bg">
        <div className="">
          <div className="bg-white dark:bg-dark-bg shadow-lg px-5 py-8 rounded-md w-full">
            <div className=""></div>
            <div>
              <div className="-mx-4 sm:-mx-8 px-4 sm:px-8 py-2 overflow-x-auto">
                <div className="inline-block w-full lg:min-w-full shadow rounded-lg overflow-hidden">
                  <table className="min-w-full leading-normal">
                    <tbody>
                      <tr>
                        <th>Week</th>
                        <th>Monday</th>
                        <th>Tuesday</th>
                        <th>Wednesday</th>
                        <th>Thursday</th>
                        <th>Friday</th>
                      </tr>
                      {weekAttendance &&
                        weekAttendance.map((week: any) => (
                          <tr key={week.weekNumber}>
                            <td
                              data-testid={week.weekNumber}
                              className="border-b border-gray-300 px-4 py-2 text-center"
                            >
                              {week.weekNumber}
                            </td>
                            {week.traineeAttendance.map((dayData: any) => (
                              <td
                                data-testid={dayData.days}
                                className="border-b border-gray-300 px-4 py-2 text-center"
                                key={dayData.days}
                              >
                                {dayData.value}
                              </td>
                            ))}
                          </tr>
                        ))}
                    </tbody>
                  </table>
                </div>
              </div>
            </div>
          </div>
          <div className="flex items-center justify-center gap-1 mt-12 mb-0">
            <button
              onClick={prevPage}
              data-testid="prev"
              className={`page flex text-white h-12 w-12 items-center justify-center border-solid cursor-pointer bg-transparent ${
                page === 1 && 'disabled'
              }`}
            >
              &larr;
            </button>
            <button
              onClick={() => setPage(1)}
              data-testid="page1"
              className={`page flex text-white h-12 w-12 items-center justify-center border-solid cursor-pointer bg-transparent ${
                page === 1 && 'disabled'
              }`}
            >
              1
            </button>
            {gaps.paginationGroup.map((el) => (
              <button
                onClick={() => setPage(el)}
                data-testid="page2"
                key={el}
                className={`page flex text-white h-12 w-12 items-center justify-center border-solid cursor-pointer bg-transparent ${
                  page === el ? 'active' : ''
                }`}
              >
                {el}
              </button>
            ))}
            <button
              onClick={() => setPage(totalPages)}
              data-testid="page3"
              className={`page flex text-white h-12 w-12 items-center justify-center border-solid cursor-pointer bg-transparent ${
                page === totalPages && 'disabled'
              }`}
            >
              {totalPages}
            </button>
            <button
              onClick={nextPage}
              data-testid="next"
              className={`page flex text-white h-12 w-12 items-center justify-center border-solid cursor-pointer bg-transparent ${
                page === totalPages && 'disabled'
              }`}
            >
              &rarr;
            </button>
          </div>
        </div>
      </div>
    </>
  );
};

export default TraineeAttendance;
