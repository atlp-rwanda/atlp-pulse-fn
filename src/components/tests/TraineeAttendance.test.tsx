/* eslint-disable */
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import TraineeAttendance from '../TraineeAttendance';


describe('Trainee Attendance page', () => {
  it('Renders the Trainee Attendance Page ', () => {
    const elem = renderer
      .create(
        <MemoryRouter>
          <TraineeAttendance />
        </MemoryRouter>,
      )
      .toJSON();
    expect(elem).toMatchSnapshot();
  });
});
