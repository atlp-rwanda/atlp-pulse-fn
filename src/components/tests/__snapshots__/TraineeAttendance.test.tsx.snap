// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Trainee Attendance page Renders the Trainee Attendance Page  1`] = `
<div
  className="bg-light-bg dark:bg-dark-frame-bg min-h-screen lg:px-8"
>
  <div
    className="px-3 md:px-8"
  >
    <div
      className="bg-white dark:bg-dark-bg shadow-lg px-5 py-8 rounded-md w-full lg:w-[80%] lg:ml-56 mt-[90px]"
    >
      <div
        className=""
      >
        <div
          className="flex ml-2 items-center justify-between"
        >
          <h2
            className="text-gray-800 dark:text-white font-semibold"
          >
            Attendance
          </h2>
        </div>
        <div
          className="flex ml-[-25px] px-7 py-2  mt-4"
        >
          <select
            className="flex bg-primary px-4 py-2 rounded-md text-white font-medium cursor-pointer"
          >
            <option>
              All phases
            </option>
            <option>
              Phase 1
            </option>
            <option>
              Phase 2
            </option>
            <option>
              Phase 3
            </option>
            <option>
              Phase 4
            </option>
            <option>
              Phase 5
            </option>
          </select>
        </div>
      </div>
      <div>
        <div
          className="-mx-4 sm:-mx-8 px-4 sm:px-8 py-2 overflow-x-auto"
        >
          <div
            className="inline-block w-full lg:min-w-full shadow rounded-lg overflow-hidden"
          >
            <table
              className="min-w-full leading-normal"
            >
              <tbody>
                <tr>
                  <th
                    className="p-6 border-b-2 border-gray-200 bg-gray-100 dark:bg-dark-tertiary text-center text-xs font-semibold text-gray-600 dark:text-white uppercase tracking-wider"
                  >
                    Sprint
                  </th>
                  <th
                    className="px-5 py-3 border-b-2 border-gray-200 bg-gray-100 dark:bg-dark-tertiary text-center text-xs font-semibold text-gray-600 dark:text-white uppercase md:table-cell sm:hidden tracking-wider"
                  >
                    Session
                  </th>
                  <th
                    className="px-5 py-3 border-b-2 border-gray-200 bg-gray-100 dark:bg-dark-tertiary text-center text-xs font-semibold text-gray-600 dark:text-white uppercase tracking-wider"
                  >
                    Record
                  </th>
                  <th
                    className="px-5 py-3 border-b-2 border-gray-200 bg-gray-100 dark:bg-dark-tertiary text-center text-xs font-semibold text-gray-600 dark:text-white uppercase tracking-wider"
                  />
                </tr>
                <tr>
                  <td
                    className="px-5 py-5 border-b border-gray-200 bg-white dark:bg-dark-bg text-sm"
                  >
                    <div
                      className="flex justify-center items-center"
                    >
                      <div
                        className=""
                      >
                        <p
                          className="text-gray-900 text-center dark:text-white whitespace-no-wrap"
                        >
                          Sprint 1
                        </p>
                      </div>
                    </div>
                  </td>
                  <td
                    className="px-5 py-5 border-b border-gray-200 bg-white dark:bg-dark-bg text-sm md:table-cell sm:hidden"
                  >
                    <p
                      className="text-gray-900  dark:text-white whitespace-no-wrap text-center "
                    >
                      Demo
                    </p>
                  </td>
                  <td
                    className="px-5 py-5 border-b border-gray-200 bg-white dark:bg-dark-bg text-sm"
                  >
                    <p
                      className="text-gray-900  dark:text-white whitespace-no-wrap text-center"
                    >
                      2
                    </p>
                  </td>
                  <td
                    className="px-5 py-5 border-b border-gray-200 bg-white dark:bg-dark-bg text-sm"
                  >
                    <a
                      href="/dashboard/trainee/attendance-details"
                      onClick={[Function]}
                    >
                      <button
                        className="px-4 py-0 rounded-md dark:text-dark-text-fill bg-primary text-white text-sm"
                      >
                        Details
                      </button>
                    </a>
                  </td>
                </tr>
                <tr>
                  <td
                    className="px-5 py-5 border-b border-gray-200 bg-gray-100 dark:bg-dark-tertiary text-sm"
                  >
                    <div
                      className="flex justify-center items-center"
                    >
                      <div
                        className=""
                      >
                        <p
                          className="text-gray-900 text-center dark:text-white whitespace-no-wrap"
                        >
                          Sprint 2
                        </p>
                      </div>
                    </div>
                  </td>
                  <td
                    className="px-5 py-5 border-b border-gray-200 bg-gray-100 dark:bg-dark-tertiary text-sm md:table-cell sm:hidden"
                  >
                    <p
                      className="text-gray-900 text-center dark:text-white whitespace-no-wrap "
                    >
                      Standup
                    </p>
                  </td>
                  <td
                    className="px-5 py-5 border-b border-gray-200 bg-gray-100 dark:bg-dark-tertiary text-sm"
                  >
                    <p
                      className="text-gray-900 text-center dark:text-white whitespace-no-wrap "
                    >
                      1
                    </p>
                  </td>
                  <td
                    className="px-5 py-5 border-b border-gray-200 bg-gray-100 dark:bg-dark-tertiary text-sm"
                  >
                    <a
                      href="/dashboard/trainee/attendance-details"
                      onClick={[Function]}
                    >
                      <button
                        className="px-4 py-0 rounded-md dark:text-dark-text-fill bg-primary text-white text-sm"
                      >
                        Details
                      </button>
                    </a>
                  </td>
                </tr>
                <tr>
                  <td
                    className="px-5 py-5 border-b border-gray-200 bg-white dark:bg-dark-bg text-sm"
                  >
                    <div
                      className="flex justify-center items-center"
                    >
                      <div
                        className=""
                      >
                        <p
                          className="text-gray-900 text-center dark:text-white whitespace-no-wrap"
                        >
                          Sprint 3
                        </p>
                      </div>
                    </div>
                  </td>
                  <td
                    className="px-5 py-5 border-b border-gray-200 bg-white dark:bg-dark-bg text-sm md:table-cell sm:hidden"
                  >
                    <p
                      className="text-gray-900 text-center dark:text-white whitespace-no-wrap"
                    >
                      Demo
                    </p>
                  </td>
                  <td
                    className="px-5 py-5 border-b border-gray-200 bg-white dark:bg-dark-bg text-sm"
                  >
                    <p
                      className="text-gray-900 text-center dark:text-white whitespace-no-wrap "
                    >
                      1
                    </p>
                  </td>
                  <td
                    className="px-5 py-5 border-b border-gray-200 bg-white dark:bg-dark-bg text-sm"
                  >
                    <a
                      href="/dashboard/trainee/attendance-details"
                      onClick={[Function]}
                    >
                      <button
                        className="px-4 py-0 rounded-md dark:text-dark-text-fill bg-primary text-white text-sm"
                      >
                        Details
                      </button>
                    </a>
                  </td>
                </tr>
                <tr>
                  <td
                    className="px-5 py-5 border-b border-gray-200 bg-gray-100 dark:bg-dark-tertiary text-sm"
                  >
                    <div
                      className="flex justify-center items-center"
                    >
                      <div
                        className=""
                      >
                        <p
                          className="text-gray-900 text-center dark:text-white whitespace-no-wrap"
                        >
                          Sprint 4
                        </p>
                      </div>
                    </div>
                  </td>
                  <td
                    className="px-5 py-5 border-b border-gray-200 bg-gray-100 dark:bg-dark-tertiary text-sm md:table-cell sm:hidden"
                  >
                    <p
                      className="text-gray-900 text-center dark:text-white whitespace-no-wrap"
                    >
                      Standup
                    </p>
                  </td>
                  <td
                    className="px-5 py-5 border-b border-gray-200 bg-gray-100 dark:bg-dark-tertiary text-sm"
                  >
                    <p
                      className="text-gray-900 text-center dark:text-white whitespace-no-wrap"
                    >
                      2
                    </p>
                  </td>
                  <td
                    className="px-5 py-5 border-b border-gray-200 bg-gray-100 dark:bg-dark-tertiary text-sm"
                  >
                    <a
                      href="/dashboard/trainee/attendance-details"
                      onClick={[Function]}
                    >
                      <button
                        className="px-4 py-0 rounded-md dark:text-dark-text-fill bg-primary text-white text-sm"
                      >
                        Details
                      </button>
                    </a>
                  </td>
                </tr>
                <tr>
                  <td
                    className="px-5 py-5 border-b border-gray-200 bg-white dark:bg-dark-bg text-sm"
                  >
                    <div
                      className="flex justify-center items-center"
                    >
                      <div
                        className=""
                      >
                        <p
                          className="text-gray-900 text-center dark:text-white whitespace-no-wrap"
                        >
                          Sprint 5
                        </p>
                      </div>
                    </div>
                  </td>
                  <td
                    className="px-5 py-5 border-b border-gray-200 bg-white dark:bg-dark-bg text-sm md:table-cell sm:hidden"
                  >
                    <p
                      className="text-gray-900 text-center dark:text-white whitespace-no-wrap"
                    >
                      Demo
                    </p>
                  </td>
                  <td
                    className="px-5 py-5 border-b border-gray-200 bg-white dark:bg-dark-bg text-sm"
                  >
                    <p
                      className="text-gray-900 text-center dark:text-white whitespace-no-wrap "
                    >
                      2
                    </p>
                  </td>
                  <td
                    className="px-5 py-5 border-b border-gray-200 bg-white dark:bg-dark-bg text-sm"
                  >
                    <a
                      href="/dashboard/trainee/attendance-details"
                      onClick={[Function]}
                    >
                      <button
                        className="px-4 py-0 rounded-md dark:text-dark-text-fill bg-primary text-white text-sm"
                      >
                        Details
                      </button>
                    </a>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
